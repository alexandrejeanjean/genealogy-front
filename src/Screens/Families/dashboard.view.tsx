import React from "react";

import ListWrapper from "../../SharedComponents/ListWrapper/ListWrapper";
import CreateItemBtn from "../../SharedComponents/CreateItemBtn/CreateItemBtn";
import ModalForm from "../../SharedComponents/ModalForm/ModalForm";
import { family } from "../../assets/imgPath";

type TFamily = {
  id: number;
  name: string;
};

type TGetFamilies = TFamily[];

type TDashboardView = {
  list: TGetFamilies;
  goTo: Function;
  deleteFamily: Function;
  setModalShow: Function;
  modalShow: boolean;
  createFamily: Function;
};

const DashboardView = ({
  list,
  goTo,
  deleteFamily,
  setModalShow,
  modalShow,
  createFamily,
}: TDashboardView) => {
  return (
    <>
      <section>
        <ListWrapper
          datas={list}
          getAvatar={() => family}
          handleClick={(familyId: number) => goTo(familyId)}
          handleDelete={(family: TFamily) => deleteFamily(family)}
        >
          <CreateItemBtn handleClick={() => setModalShow(true)} />
        </ListWrapper>
      </section>
      <ModalForm
        title="Family"
        show={modalShow}
        inputs={[
          { name: "name", placeholder: "Ex: Armstrong", inputType: "text" },
        ]}
        onHide={() => setModalShow(false)}
        submit={createFamily}
      />
    </>
  );
};

export default DashboardView;
