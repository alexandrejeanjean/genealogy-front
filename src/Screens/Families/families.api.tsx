import FamilyModel from "../../models/family.model";
import { ErrorHandler } from "../../helpers";
import apiClient from "../../api";

/**
 * get list of families for the given user
 * @returns {Array<Family>} - return list of families
 */

export async function apiGetFamilies(setToastVisible: Function) {
  try {
    const response = await apiClient.get("/families");
    const families = response.data;
    if (families) {
      return families.map((Family: any) => new FamilyModel(Family));
    }
  } catch (err) {
    ErrorHandler(err, setToastVisible, "We can't access to your families.");
  }
}

/**
 * post new family for the given user
 * @param {Function} setToastVisible - function to load modal
 * @param {Object} family - family datas
 * @returns return 201 if new family created successfully
 */

export async function apiCreateFamily(
  setToastVisible: Function,
  family: { name: string }
) {
  try {
    const response = await apiClient.post("/families", { name: family.name });
    if (response) {
      return response;
    }
  } catch (err) {
    ErrorHandler(err, setToastVisible, "We can't create your family.");
  }
}

/**
 * delete a family for the given user by id
 * @param {Function} setToastVisible - function to load modal
 * @param {number}  - "family id to delete"
 * @returns return 204 if family deleted successfully
 */

export async function apiDeleteFamily(
  setToastVisible: Function,
  familyId: number
) {
  try {
    const response = await apiClient.delete(`/families/${familyId}`);
    if (response) {
      return response;
    }
  } catch (err) {
    ErrorHandler(err, setToastVisible, "We can't delete this family.");
  }
}
