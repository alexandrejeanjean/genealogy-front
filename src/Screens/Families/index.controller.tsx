import React, { useEffect, useState, useCallback } from "react";
import { useHistory } from "react-router-dom";
import { withToast } from "../../store/ToastProvider";

import {
  apiGetFamilies,
  apiCreateFamily,
  apiDeleteFamily,
} from "./families.api";

import Main from "../Theme/index.view";
import DashboardView from "./dashboard.view";

import "./dashboard.scss";

type TFamily = {
  id: number;
  name: string;
};

type TGetFamilies = TFamily[];

type TDashboard = {
  setToastVisible: Function;
};

const Dashboard = ({ setToastVisible }: TDashboard) => {
  const [list, setList] = useState<TGetFamilies>([]);
  const [modalShow, setModalShow] = useState<boolean>(false);

  const history = useHistory();

  const getFamilies = useCallback(async () => {
    apiGetFamilies(setToastVisible)
      .then((res) => setList(res))
      .catch((err) => console.log(err));
  }, [setToastVisible]);

  // Get datas on component did mount --------------------------
  useEffect(() => {
    getFamilies();
  }, [getFamilies]);

  // Add family  ---------------------------------------
  const createFamily = async (family: { name: string }) => {
    apiCreateFamily(setToastVisible, family)
      .then(() => getFamilies())
      .catch((err) => console.log(err));
  };

  // Delete family  ---------------------------------------
  const deleteFamily = async (family: TFamily) => {
    const familyId: number = family.id;
    apiDeleteFamily(setToastVisible, familyId)
      .then(() => getFamilies())
      .catch((err) => console.log(err));
  };

  const goTo = (familyId: number) => {
    return history.push({
      pathname: "/family",
      state: { datas: familyId },
    });
  };

  return (
    <Main pageTitle="Dashboard">
      {console.log("LIST ::", list)}
      <DashboardView
        list={list}
        goTo={(familyId: number) => goTo(familyId)}
        deleteFamily={(family: TFamily) => deleteFamily(family)}
        setModalShow={(value: boolean) => setModalShow(value)}
        modalShow={modalShow}
        createFamily={createFamily}
      />
    </Main>
  );
};

export default withToast(Dashboard);
