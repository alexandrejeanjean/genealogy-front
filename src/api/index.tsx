import axios from "axios";

export function setAuthorization(Authorization: any) {
  apiClient.defaults.headers.common.Authorization = Authorization;
}

const apiClient = axios.create({
  baseURL: process.env.REACT_APP_API_URL + "/api",
  responseType: "json",
  headers: {
    "Content-Type": "application/json",
  },
});

export default apiClient;
