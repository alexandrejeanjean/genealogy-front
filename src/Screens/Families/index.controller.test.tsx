import React from "react";
import { shallow, mount } from "enzyme";
import FamilyModel from "../../models/family.model";
import Families from "./index.controller";
import { apiGetFamilies } from "./families.api";
const axios = require("axios");
jest.mock("axios");

const mockedApiClient = {
  apiGetFamilies: jest.fn(),
};

const rawFamily = [
  {
    id: 1,
    name: "Armstrong",
    userId: 1,
  },
  {
    id: 2,
    name: "Jazz family",
    userId: 2,
  },
];

afterAll(() => jest.resetAllMocks());

const FamiliesProps = {
  setToastVisible: jest.fn(),
};

describe("Families", () => {
  it("should render correctly", () => {
    const Component = shallow(<Families {...FamiliesProps} />);
    expect(Component).toMatchSnapshot();
  });
});

describe("Family API", () => {
  it("should return the list of families", async () => {
    mockedApiClient.apiGetFamilies.mockResolvedValueOnce(rawFamily);
    const families = await mockedApiClient.apiGetFamilies(
      FamiliesProps.setToastVisible
    );
    expect(families).toEqual(rawFamily.map((family) => family));
  });
});
