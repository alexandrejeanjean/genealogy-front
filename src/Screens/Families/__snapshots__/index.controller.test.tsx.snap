// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Families should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
